### PR TITLE
Avoid hashing length for ty::List

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
r? @ghost -- let's gather perf stats first, see also https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/ty.3A.3AList.20StableHash

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
```
error: symbol `_ZN100_$LT$core..iter..adapters..fuse..Fuse$LT$I$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$8try_fold17h8a5af79663957c6eE` is already defined
   --> src/libcore/iter/adapters/fuse.rs:93:5
    |
93  | /     fn try_fold<Acc, Fold, R>(&mut self, acc: Acc, fold: Fold) -> R
94  | |     where
95  | |         Self: Sized,
96  | |         Fold: FnMut(Acc, Self::Item) -> R,
...   |
99  | |         FuseImpl::try_fold(self, acc, fold)
100 | |     }
    | |_____^

error: aborting due to previous error
```

Hm, this implies that maybe ty::List can be subsliced? But then the normal Hash impl and PartialEq are wrong...

File: src/librustc_middle/ich/impls_ty.rs [hunk 1]
```diff
@@ -17,12 +17,12 @@ where
...
     fn hash_stable(&self, hcx: &mut StableHashingContext<'a>, hasher: &mut StableHasher) {
         thread_local! {
-            static CACHE: RefCell<FxHashMap<(usize, usize), Fingerprint>> =
+            static CACHE: RefCell<FxHashMap<usize, Fingerprint>> =
                 RefCell::new(Default::default());
         }
 
         let hash = CACHE.with(|cache| {
-            let key = (self.as_ptr() as usize, self.len());
+            let key = self as *const _ as usize;
```
Comment (matthewjasper) [inline src/librustc_middle/ich/impls_ty.rs:25]: `self` is &&'tcx List<T>` here, so this is isn't the address of the interned pointer.

Comment (Mark-Simulacrum) [reply src/librustc_middle/ich/impls_ty.rs:25]: No wonder this didn't work :)

**Comment by Mark-Simulacrum** [comment]
Seems like mostly not any difference, maybe some very slight regressions? That's somewhat surprising to me -- I would've expected us to be hashing lists more -- but I guess maybe we don't do it quite that often for it to contribute any significant percentage of instruction counts.

**Pull request closed without merging**